src/btree/bt_sync_obsolete.c

```diff
@@ -291,3 +290,324 @@ __wt_sync_obsolete_cleanup(WT_SESSION_IMPL *session, WT_REF *parent)
+     */
+    if (ref->state == WT_REF_DELETED) {
+        *skipp = true;
+        return (0);
+    }
+
+    /* If the page is in-memory, we want to look at it. */
+    if (ref->state != WT_REF_DISK)
+        return (0);
+
+    /*
+     * Reading any page that is not in the cache will increase the cache size. Perform a set of
+     * checks to verify the cache can handle it.
+     */
+    if (__wt_cache_aggressive(session) || __wt_cache_full(session) || __wt_cache_stuck(session) ||
+      __wt_eviction_needed(session, false, false, NULL)) {
+        *skipp = true;
+        return (0);
+    }
+
+    /* Don't read pages into cache during startup or shutdown phase. */
+    if (F_ISSET(S2C(session), WT_CONN_RECOVERING | WT_CONN_CLOSING_CHECKPOINT)) {
+        *skipp = true;
+        return (0);
+    }
+
+    /*
+     * Ignore the pages with no on-disk address. It is possible that a page with deleted state may
+     * not have an on-disk address.
+     */
+    if (!__wt_ref_addr_copy(session, ref, &addr))
+        return (0);
+
+    /*
+     * The checkpoint cleanup fast deletes the obsolete leaf page by marking it as deleted
+     * in the internal page. To achieve this,
+     *
+     * 1. Checkpoint has to read all the internal pages that have obsolete leaf pages.
+     *    To limit the reading of number of internal pages, the aggregated stop durable timestamp
+     *    is checked except when the table is logged. Logged tables do not use timestamps.
+     *
+     * 2. Obsolete leaf pages with overflow keys/values cannot be fast deleted to free
+     *    the overflow blocks. Read the page into cache and mark it dirty to remove the
+     *    overflow blocks during reconciliation.
+     *
+     * FIXME: Read internal pages from non-logged tables when the remove/truncate
+     * operation is performed using no timestamp.
+     */
+
+    if (addr.type == WT_ADDR_LEAF_NO ||
+      (addr.ta.newest_stop_durable_ts == WT_TS_NONE &&
+        (F_ISSET(S2C(session), WT_CONN_CKPT_CLEANUP_SKIP_INT) ||
+          !F_ISSET(S2BT(session), WT_BTREE_LOGGED)))) {
+        __wt_verbose_debug2(
+          session, WT_VERB_CHECKPOINT_CLEANUP, "%p: page walk skipped", (void *)ref);
+        WT_STAT_CONN_DATA_INCR(session, checkpoint_cleanup_pages_walk_skipped);
+        *skipp = true;
+    }
+    return (0);
+}
+
+/*
+ * __checkpoint_cleanup_walk_btree --
+ *     Check and perform checkpoint cleanup on the uri.
+ */
+static int
+__checkpoint_cleanup_walk_btree(WT_SESSION_IMPL *session, const char *uri)
+{
+    WT_BTREE *btree;
+    WT_DECL_RET;
+    WT_REF *ref;
+    uint32_t flags;
+
+    ref = NULL;
+    flags = WT_READ_NO_EVICT;
+
+    /*
+     * To reduce the impact of checkpoint cleanup on the running database, it operates only on the
+     * dhandles that are already opened.
+     */
+    WT_WITHOUT_DHANDLE(session,
+      WT_WITH_HANDLE_LIST_READ_LOCK(session, (ret = __wt_conn_dhandle_find(session, uri, NULL))));
+    if (ret == WT_NOTFOUND)
```

Comment:
What are the impacts if that dhandle is never re opened?

Reply:
Normally, checkpoints work only on the dhandles that are modified; as we separated checkpoint cleanup from checkpoints, it should work on all the dhandles present in the cache. If it doesn't work, any obsolete pages on those dhandles stay as they are on the disk.